src/options/template.rs

```diff
@@ -42,48 +44,86 @@ impl TemplateOpt {
-    let mut f = std::fs::File::create(&tarball_path)?;
-    eprintln!("Downloading from {}", &download_url);
-    let response_bytes = client
-        .get(download_url)
-        .header(reqwest::header::USER_AGENT, "rover-client")
-        .header(reqwest::header::ACCEPT, "application/octet-stream")
-        .send()
-        .await?
-        .error_for_status()?
-        .bytes()
-        .await?;
-    f.write_all(&response_bytes[..])?;
-    f.sync_all()?;
-    let f = std::fs::File::open(&tarball_path)?;
-    let tar = flate2::read::GzDecoder::new(f);
-    let mut archive = tar::Archive::new(tar);
-    archive
-        .unpack(template_path)
-        .with_context(|| format!("could not unpack tarball to '{}'", &template_path))?;
-
-    // The unpacked tar will be nested in another folder
-    let extra_dir_name = Fs::get_dir_entries(template_path)?.find(|_| true);
-    if let Some(Ok(extra_dir_name)) = extra_dir_name {
-        // For this reason, we must copy the contents of the folder, then delete it
-        Fs::copy_dir_all(extra_dir_name.path(), template_path)?;
-
-        // Delete old unpacked zip
-        Fs::remove_dir_all(extra_dir_name.path())?;
+#[derive(Debug)]
+pub struct TemplateFetcher {
+    request_service: ReqwestService,
+}
+
+#[derive(Debug)]
+pub struct TemplateProject {
+    contents: Vec<u8>,
+}
+
+impl TemplateFetcher {
+    pub fn new(request_service: ReqwestService) -> Self {
+        Self { request_service }
+    }
+
+    pub async fn call(&mut self, download_url: Uri) -> RoverResult<TemplateProject> {
+        println!("Downloading from {}", &download_url);
+        println!();
+        let req = http::Request::builder()
+            .method(http::Method::GET)
+            .header(reqwest::header::ACCEPT, "application/octet-stream")
+            .header(reqwest::header::USER_AGENT, "rover-client")
+            .uri(download_url)
+            .body(Full::default())?;
+
+        let service = self.request_service.ready().await?;
+        let res = service.call(req).await?;
+        let res = res.body().to_vec();
+
+        if res.is_empty() {
+            return Err(RoverError::new(anyhow!("No template found")));
+        }
+
+        Ok(TemplateProject { contents: res })
+    }
+}
+
+impl TemplateProject {
+    pub fn write_template(&self, template_path: &Utf8PathBuf) -> RoverResult<()> {
+        let cursor = Cursor::new(&self.contents);
+        let tar = flate2::read::GzDecoder::new(cursor);
+        let mut archive = tar::Archive::new(tar);
+
+        archive.unpack(template_path)?;
+
+        let extra_dir_name = Fs::get_dir_entries(template_path)?.find(|_| true);
+        if let Some(Ok(extra_dir_name)) = extra_dir_name {
+            Fs::copy_dir_all(extra_dir_name.path(), template_path)?;
+            Fs::remove_dir_all(extra_dir_name.path())?;
+        }
+
+        Ok(())
     }
 
-    Ok(())
+    // this is only used by init behind feature flag
+    #[allow(dead_code)]
```

Comment:
```suggestion
    #[cfg_attr(not(feature = "init"), allow(dead_code))]
```
That way, if the use of `list_files` were removed, the compiler would tell you.